crates/parcel/src/requests/target_request.rs

```diff
@@ -0,0 +1,1107 @@
+    BuiltInTarget {
+      descriptor: descriptor.unwrap_or_else(|| {
+        BuiltInTargetDescriptor::TargetDescriptor(TargetDescriptor {
+          context: Some(EnvironmentContext::Node),
+          ..self.builtin_target_descriptor()
+        })
+      }),
+      dist,
+      extensions: vec!["ts"],
+      name: "types",
+    }
+  }
+
+  fn default_dist_dir(&self, package_path: &Path) -> PathBuf {
+    package_path
+      .parent()
+      .unwrap_or_else(|| &package_path)
+      .join("dist")
+  }
+
+  fn fallback_output_format(&self, context: EnvironmentContext) -> OutputFormat {
+    match context {
+      EnvironmentContext::Node => OutputFormat::CommonJS,
+      EnvironmentContext::ElectronMain => OutputFormat::CommonJS,
+      EnvironmentContext::ElectronRenderer => OutputFormat::CommonJS,
+      _ => OutputFormat::Global,
+    }
+  }
+
+  fn infer_environment_context(&self, package_json: &PackageJson) -> EnvironmentContext {
+    // If there is a separate `browser` target, or an `engines.node` field but no browser
+    // targets, then the target refers to node, otherwise browser.
+    if package_json.browser.is_some() || package_json.targets.browser.is_some() {
+      if package_json
+        .engines
+        .as_ref()
+        .is_some_and(|e| e.node.is_some() && e.browsers.is_empty())
+      {
+        return EnvironmentContext::Node;
+      } else {
+        return EnvironmentContext::Browser;
+      }
+    }
+
+    if package_json
+      .engines
+      .as_ref()
+      .is_some_and(|e| e.node.is_some())
+    {
+      return EnvironmentContext::Node;
+    }
+
+    EnvironmentContext::Browser
+  }
+
+  fn infer_output_format(
+    &self,
+    module_format: &Option<ModuleFormat>,
+    target: &TargetDescriptor,
+  ) -> Result<Option<OutputFormat>, anyhow::Error> {
+    let ext = target
+      .dist_entry
+      .as_ref()
+      .and_then(|e| e.extension())
+      .unwrap_or_default()
+      .to_str();
+
+    let inferred_output_format = match ext {
+      Some("cjs") => Some(OutputFormat::CommonJS),
+      Some("mjs") => Some(OutputFormat::EsModule),
+      Some("js") => module_format.as_ref().and_then(|format| match format {
+        ModuleFormat::CommonJS => Some(OutputFormat::CommonJS),
+        ModuleFormat::Module => Some(OutputFormat::EsModule),
+      }),
+      _ => None,
+    };
+
+    if let Some(inferred_output_format) = inferred_output_format {
+      if let Some(output_format) = target.output_format {
+        if output_format != inferred_output_format {
+          return Err(anyhow!(
+            "Declared output format {} does not match expected output format {}",
+            output_format,
+            inferred_output_format
+          ));
+        }
+      }
+    }
+
+    Ok(inferred_output_format)
+  }
+
+  fn load_package_json(&self) -> Result<(PathBuf, PackageJson), anyhow::Error> {
+    // TODO Invalidations
+    let (package_path, mut package_json) = self.config.load_package_json_config::<PackageJson>()?;
```

Comment:
IMO anything that does any reading should be coming either as an input to the request or it should be done through that context parameter.

That is because we will be able to implement read-tracking through the context parameter to not have to track invalidations manually

Reply:
Although I agree in premise, I think we should do a bit of sparring on this before proceeding and apply it to all requests like I said in another comment